contentcuration/automation/utils/appnexus/base.py

```diff
@@ -1,46 +1,183 @@
 
 
 class BackendRequest(object):
-    """ Class that should be inherited by specific backend for its requests"""
-    pass
+    """ Class that holds the request information for the backend """
+    def __init__(
+        self,
+        method,
+        path,
+        params=None,
+        data=None,
+        json=None,
+        headers=None,
+        timeout=(5, 100),
+        **kwargs
+    ):
+        self.method = method
+        self.path = path
+        self.params = params
+        self.data = data
+        self.json = json
+        self.headers = headers
+        self.timeout = timeout
+        for key, value in kwargs.items():
+            setattr(self, key, value)
 
 
 class BackendResponse(object):
     """ Class that should be inherited by specific backend for its responses"""
-    pass
+    def __init__(self, **kwargs):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
 
 
 class Backend(ABC):
     """ An abstract base class for backend interfaces that also implements the singleton pattern """
     _instance = None
-
-    def __new__(class_, *args, **kwargs):
-        if not isinstance(class_._instance, class_):
-            class_._instance = object.__new__(class_, *args, **kwargs)
-        return class_._instance
-
+    session = None
+    base_url = None
+    connect_endpoint = None
+    max_retries=1
+    backoff_factor=0.3
+
+    def __new__(cls, *args, **kwargs):
+        if not isinstance(cls._instance, cls):
+            cls._instance = object.__new__(cls)
+        return cls._instance
+
+    def __init__(
+        self,
+        url_prefix="",
+    ):
+        self.url_prefix = url_prefix
+        if not self.session:
+            self._setup_session()
+
+    def _setup_session(self):
+        self.session = SessionWithMaxConnectionAge()
+
+        retry = Retry(
+            total=self.max_retries,
+            backoff_factor=self.backoff_factor,
+        )
+        adapter = HTTPAdapter(max_retries=retry)
+
+        self.session.mount("https://", adapter)
+        self.session.mount("http://", adapter)
+
+    def _construct_full_url(self, path):
+        """This method combine base_url, url_prefix, and path in that order, removing any trailing and leading slashes."""
+        url_array = []
+        if self.base_url:
+            url_array.append(self.base_url.rstrip("/"))
+        if self.url_prefix:
+            url_array.append(self.url_prefix.rstrip("/").lstrip("/"))
+        if path:
+            url_array.append(path.lstrip("/"))
+        return "/".join(url_array)
+
+    def _make_request(self, request):
+        url = self._construct_full_url(request.path)
+        try:
+            return self.session.request(
+                request.method,
+                url,
+                params=request.params,
+                data=request.data,
+                headers=request.headers,
+                json=request.json,
+                timeout=request.timeout,
+            )
```

Comment:
Sorry for the piecemeal review-- every time I look I see something different. I think one thing that may be missing here is the `response.raise_for_status` ([Response.raise_for_status()](https://3.python-requests.org/api/#requests.Response.raise_for_status)). Using that would likely negate the need for checking the status code in `connect`

Reply:
Thanks! I have added the call to `raise_for_status`.